### PR TITLE
feat: Implement required traits for `ChunkType`

### PR DESCRIPTION
**Pull request by iandrc**
Refs:
- https://picklenerd.github.io/pngme_book/chapter_1.html

### CODE REVIEW

File: src/chunk_type.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+#[derive(PartialEq, Eq, Debug)]
+struct ChunkType {
+    bytes: [u8; 4],
+}
+
+impl TryFrom<[u8; 4]> for ChunkType {
+    type Error = &'static str;
+
+    fn try_from(value: [u8; 4]) -> Result<Self, Self::Error> {
+        if value.len() != 4 {
```
Comment (Mart-Bogdan) [inline src/chunk_type.rs:14]: This check is meaningless, as we are using array here, not slice. 

Arrays have fixed length.

File: src/chunk_type.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            Ok(ChunkType { bytes: value })
+        }
+    }
+}
+
+impl FromStr for ChunkType {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() != 4 {
```
Comment (Mart-Bogdan) [inline src/chunk_type.rs:26]: for example, her it is ok

File: src/chunk_type.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            Err("The string should have exactly 4 characters")
+        } else {
+            let mut bytes: [u8; 4] = [0; 4];
+
+            for (index, c) in "RuSt".chars().enumerate() {
```
Comment (Mart-Bogdan) [inline src/chunk_type.rs:31]: If we work with utf-8 strings beyound ASCII charset, then most chars would be > 256. Actually in rust char is u32.

I'd suggest just get u8's from string and encode them directly.

Also s.len() is not checking chars, but bytes in Rust.

**Review by Mart-Bogdan — COMMENTED** [review]
Some minor comments for code.

File: src/chunk_type.rs (hunk 1 above)
Comment (iandrc) [reply src/chunk_type.rs:14]: Does it mean Rust will guarantee that the incoming param will always have the length of 4?

Comment (Mart-Bogdan) [reply src/chunk_type.rs:14]: yes. It's statically checked,

But it won't guarantee that all indexes are filled with meaningful data.

That's up to caller.

File: src/chunk_type.rs (hunk 3 above)
Comment (iandrc) [reply src/chunk_type.rs]: Sorry, I didn't get the second part. What do you mean by saying `"encode them directly"`?

Thanks, I'll try to memorize it :)

Comment (Mart-Bogdan) [reply src/chunk_type.rs]: I mean jsut store u8's. Not funn chars. Then, hhen you extract text from PNG you could jsut put back taht u8's and assemble string.

File: src/chunk_type.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() != 4 {
+            Err("The string should have exactly 4 characters")
+        } else {
+            let mut bytes: [u8; 4] = [0; 4];
+
+            for (index, c) in s.chars().enumerate() {
+                bytes[index] = c as u8;
+            }
```
Comment (TheBestTvarynka) [inline src/chunk_type.rs:33]: make it simpler:
```rust
bytes[0..4].copy_from_slice(s.as_bytes());
```

File: src/chunk_type.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+impl TryFrom<[u8; 4]> for ChunkType {
+    type Error = &'static str;
+
+    fn try_from(value: [u8; 4]) -> Result<Self, Self::Error> {
+        if value.len() != 4 {
+            Err("The chunk's length should be exactly 4")
+        } else {
+            Ok(ChunkType { bytes: value })
```
Comment (TheBestTvarynka) [inline src/chunk_type.rs:17]: Your `TryFrom` and `FromStr` implementations are not complete. You are not checking for the actual chunk type value. We have a concrete list of allowed chunk types in the PNG format. For example, chunk type "code" does not exist. You should handle check such situations

File: src/chunk_type.rs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::convert::TryFrom;
+    use std::str::FromStr;
+
+    #[test]
+    pub fn test_chunk_type_from_bytes() {
+        let expected = [82, 117, 83, 116];
```
Comment (TheBestTvarynka) [inline src/chunk_type.rs:77]: Please, use valid chunk types in your tests

**Review by TheBestTvarynka — CHANGES_REQUESTED** [review]
It's good but I have a few comments

File: src/chunk_type.rs (hunk 6 above)
Comment (TheBestTvarynka) [reply src/chunk_type.rs:77]: Also, you may add readable chunk type value in the comment above of the array. Example:
```rust
// IHDR
let expected = [73, 72, 68, 82];
```

Comment (iandrc) [reply src/chunk_type.rs:77]: Thanks for the comment, yet it's the unit tests provided by the course designer :)

File: src/chunk_type.rs (hunk 4 above)
Comment (iandrc) [reply src/chunk_type.rs:33]: Wow, thanks. Looks much better

File: src/chunk_type.rs (hunk 3 above)
Comment (iandrc) [reply src/chunk_type.rs]: Refactored with @TheBestTvarynka 's suggestion :)

File: src/chunk_type.rs (hunk 5 above)
Comment (iandrc) [reply src/chunk_type.rs:17]: It's planned for the following PR :)

**Pull request merged**